design/automatic-network-configuration.md

```diff
@@ -0,0 +1,355 @@
+    SwitchPort configuration is completed.
+- Network CRD:
+  - Contains the network configuration of the cluster and the references
+    referring to which interfaces are currently using this network
+    configuration.
+- Network controller:
+  - Monitors Network CRs, when there is a new interface using the configuration
+    defined by a Network CR, changes the corresponding SwitchPort CR vlan
+    configuration.
+- Interface CRD:
+  - Interface CR represents a network interface, this can be a NIC in a bare
+    metal machine. It contains device information, LLDP information including
+    the connected switch’s mac address and the switch portID of the link, and
+    the reference referring to the device which this interface is connected to.
+- Interface controller:
+  - Monitors Interface CRs. When there is a new Interface CR, find the
+    corresponding SwitchPort CR according to the provided information and store
+    the reference of that SwitchPort CR in the Interface CR.
+- CAPM3:
+  - Gets the LLDP information of all the interfaces in the host. If there is no
+    corresponding Interface CR, create one.
+  - If a host has multiple interfaces, selects one for network configuration.
+  - Queries the configuration status of the SwitchPort CR to ensure that the
+    configuration is successful before provisioning.
+
+The main implementation process is described below:
+- After creating a BMH, the administrator puts an interface into the
+  provisioning network.
+
+  ![automatically configure 1](images/automatically_configure_1.png)
+
+- After an Interface is put into the provisioning network, the LLDP information
+  of other interfaces can be obtained and network configuration can be
+  performed.
+
+  ![automatically configure 2](images/automatically_configure_2.png)
+
+The workflow of a basic user story:
+- Administrator:
+  - Creates a Network resource used as provisioning network.
+  - Creates BMH CRs. Need to know the following information about BMH:
+    - Info of the interface which will be connected to the provisioning
+      network.
+    - Info of the switch port connected to this interface.
+  - Creates an Interface CR for each BMH based on the above information.
+    - The Interface controller monitors the Interface CR, finds the
+      corresponding SwitchPort CR according to the switch port information, and
+      fills it in `.spec.neighborRef` in the Interface CR.
+  - Adds the Interface CR information to `.spec.interfaces` in the provisioning
+    Network CR.
+    - The Network controller watches Network CRs and for each interfaces in its
+      `.spec.interfaces`, the controller will modify the corresponding
+      SwitchPort CR to match the network configuration.
+    - The Switch controller will configure the switch based on the SwitchPort
+      CR.
+    - After the switch configuration is completed, the BMH is connected to the
+      provisioning network. Ironic will perform introspection and update the
```

Comment:
The baremetal-operator is going to try to perform inspection as soon as it has BMC credentials for the host, which happens up on line 116. If the host has to be attached to the provisioning network before that inspection process will work, then the admin is going to have to do that very early in the process, and the tenant won't get to pick the network.